src/raft/recv_install_snapshot.c

```diff
@@ -413,30 +424,42 @@ static void follower_work_done(struct work *w)
 	follower_tick(follower, M_WORK_DONE);
 }
 
-static void to_init(struct timeout *to)
+static void rpc_to_cb(uv_timer_t *handle)
 {
-	sm_init(&to->sm, to_sm_invariant, NULL, to_sm_conf, "to", TO_INIT);
+	struct timeout *to = CONTAINER_OF(handle, struct timeout, handle);
+	struct rpc *rpc = CONTAINER_OF(to, struct rpc, timeout);
+	struct leader *leader = CONTAINER_OF(rpc, struct leader, rpc);
+	sm_move(&to->sm, TO_EXPIRED);
+	sm_move(&rpc->sm, RPC_TIMEDOUT);
+	leader_tick(leader, M_TIMEOUT);
 }
 
-static void leader_to_cb(struct timeout *t, int rc)
+static void leader_to_cb(uv_timer_t *handle)
 {
-	(void)rc;
-	struct leader *leader = CONTAINER_OF(t, struct leader, timeout);
-	sm_move(&t->sm, TO_EXPIRED);
+	struct timeout *to = CONTAINER_OF(handle, struct timeout, handle);
+	struct leader *leader = CONTAINER_OF(to, struct leader, timeout);
+	sm_move(&to->sm, TO_EXPIRED);
 	leader_tick(leader, M_TIMEOUT);
 }
 
-static void to_start(struct timeout *to, unsigned delay, to_cb_op to_cb)
+static void leader_to_start(struct leader *leader,
+		struct timeout *to,
+		unsigned delay,
+		to_cb_op to_cb)
 {
-	(void)delay;
-	struct leader *leader = CONTAINER_OF(to, struct leader, timeout);
-
-	to_init(to);
-	leader->ops->to_start(&leader->timeout, 10000, to_cb);
+	leader->ops->to_init(to);
+	sm_init(&to->sm, to_sm_invariant, NULL, to_sm_conf, "to", TO_INIT);
+	leader->ops->to_start(to, delay, to_cb);
 	sm_relate(&leader->sm, &to->sm);
 	sm_move(&to->sm, TO_STARTED);
 }
 
+static void leader_to_fini(struct leader *leader, struct timeout *to)
```

Comment:
Q: Is it better to put a pointer to `v->ops()` into `timeout`-struct itself?

Reply:
Q: is it better to rename `leader_to_fini()` to `*_cancel()`?

Reply:
Agree with the second one. For the first, I am not sure I know what you mean:
* If what you mean is to include `leader->ops` inside the timeout struct, then I hesitant to do it because then the timeout will have primitives like `send_message`, `work_queue`, etc.
* If what you mean is to create another struct of operations for timeouts (`start`, `init`, `stop`) and "pass" the functions from `leader->ops` on creation then that will make timeout more general and allow us to reuse it somewhere else.

Reply:
As discussed offline, the comment was about option 1 above. It had some challenges that required more code making the benefits less apparent. We agreed to do the changes in future iterations when we need them.